### PR TITLE
Add QgsSingleGeometryCheck

### PR DESCRIPTION
**Pull request by m-kuhn**
This class is a new subclass of QgsGeometryCheck that implements an interface that works on isolated geometries.
It has no dependency on a context, apart from a configuration that can be passed in. This makes it suitable
for checks that only work on in-memory information like self-intersection and other isValid checks.

For reference the classes QgsGeometrySelfIntersectionCheck, QgsGeometryMultipartCheck, QgsGeometryTypeCheck
and QgsGeometrySelfContactCheck have been ported to this interface.

### CODE REVIEW

File: src/analysis/vector/geometry_checker/qgsgeometryselfcontactcheck.cpp [hunk 1]
```diff
@@ -17,70 +17,67 @@
...
 
-void QgsGeometrySelfContactCheck::collectErrors( QList<QgsGeometryCheckError *> &errors, QStringList &/*messages*/, QAtomicInt *progressCounter, const QMap<QString, QgsFeatureIds> &ids ) const
+QList<QgsSingleGeometryCheckError *> QgsGeometrySelfContactCheck::processGeometry( const QgsGeometry &geometry, const QVariantMap &configuration ) const
 {
-  QMap<QString, QgsFeatureIds> featureIds = ids.isEmpty() ? allLayerFeatureIds() : ids;
-  QgsGeometryCheckerUtils::LayerFeatures layerFeatures( mContext->featurePools, featureIds, mCompatibleGeometryTypes, progressCounter, mContext );
-  for ( const QgsGeometryCheckerUtils::LayerFeature &layerFeature : layerFeatures )
+  QList<QgsSingleGeometryCheckError *> errors;
+  const QgsAbstractGeometry *geom = geometry.constGet();
+  for ( int iPart = 0, nParts = geom->partCount(); iPart < nParts; ++iPart )
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgsgeometryselfcontactcheck.cpp]: Off topic: we really need proper part/ring iterators in QgsGeometry

File: src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+    /**
+     * The vertex id of the error. May be invalid depending on the check.
+     *
+     * \since QGIS 3.4
+     */
+    QgsVertexId vertexId() const;
+
+  protected:
+    const QgsSingleGeometryCheck *mCheck;
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h]: = nullptr

File: src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    QgsSingleGeometryCheckError( const QgsSingleGeometryCheck *check, const QgsGeometry &geometry, const QgsPoint &errorLocation, const QgsVertexId &vertexId = QgsVertexId() )
+      : mCheck( check )
+      , mGeometry( geometry )
+      , mErrorLocation( errorLocation )
+      , mVertexId( vertexId )
+    {}
+
+    virtual ~QgsSingleGeometryCheckError() = default;
+
+    virtual void update( const QgsSingleGeometryCheckError *other );
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h]: Dox

File: src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+     * The vertex id of the error. May be invalid depending on the check.
+     *
+     * \since QGIS 3.4
+     */
+    QgsVertexId vertexId() const;
+
+  protected:
+    const QgsSingleGeometryCheck *mCheck;
+    QgsGeometry mGeometry;
+    QgsPoint mErrorLocation;
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h]: Should this support multiple locations? Eg self intersection multiple times

File: src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+    /**
+     * The underlying single error.
+     */
+    QgsSingleGeometryCheckError *singleError() const;
+
+    bool handleChanges( const QgsGeometryCheck::Changes &changes ) override;
+
+  private:
+    QgsSingleGeometryCheckError *mError;
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h]: = nullptr

File: src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h (hunk 4 above)
Comment (m-kuhn) [reply src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h]: I was more wondering if it should support polygons (overlap etc) it's a pure port of what's existing

File: src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h (hunk 3 above)
Comment (m-kuhn) [reply src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h]: Still trying to figure out what that actually does :D

File: src/analysis/vector/geometry_checker/qgsgeometryselfcontactcheck.cpp (hunk 1 above)
Comment (m-kuhn) [reply src/analysis/vector/geometry_checker/qgsgeometryselfcontactcheck.cpp]: I think if you look at a diff with whitespace removed you won't notice :P

**Comment by m-kuhn** [comment]
> It has no dependency on a context, apart from a configuration that can be passed in

That's actually a lie. E.g. the geometry type check still depends on the type variable passed in in the constructor.

File: src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h (hunk 3 above)
Comment (manisandro) [reply src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h]: @m-kuhn When an error is fixed, affected geometries are re-scanned, and the new list of found errors is compared to the previously found errors. If a new error is equal (or close match) to an old one, the old error is updated (i.e. with new coordinates or indices, as a consequence of the changes performed).

File: src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h (hunk 4 above)
Comment (nyalldawson) [reply src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h]: Yeah - that's a very good point. Maybe this should just be a QgsGeometry for error location, so multipoint, lines, polygons, etc are supported

File: src/analysis/vector/geometry_checker/qgsgeometryselfcontactcheck.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/analysis/vector/geometry_checker/qgsgeometryselfcontactcheck.cpp]: Come on - tell me you wouldn't prefer:

     for ( const auto part : geom.parts() )
     {
        for ( const auto ring: part.rings() )
           ...

Get on it! Boy scout that code! ;)

**Comment by nyalldawson** [comment]
Bonus points if you find a clang-tidy fixit which moves ALL the implementations to CPP ;)

File: src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h (hunk 3 above)
Comment (m-kuhn) [reply src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h]: Thanks @manisandro do the docs look good to you?

Comment (manisandro) [reply src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h]: Yep looks good

**Pull request merged**